src/worker/users.js

```diff
@@ -77,37 +85,77 @@ class Users {
             return null;
         }
 
-        if (checkPass && !await user.checkPassword(password)) {
+        if (!isUserToken && !await user.checkPassword(password)) {
             return null;
         }
 
+        if (isUserToken) {
+            this.updateUserTokenAccess(user.id, password, userHost);
+        }
+
         return user;
     }
 
-    async authUserToken(token) {
+    async authUserToken(token, userHost) {
         let user = this.db.dbUsers('users')
             .innerJoin('user_tokens', 'users.id', 'user_tokens.user_id')
             .where('user_tokens.token', token)
             .first()
             .then(this.db.factories.User.fromDbResult);
 
         if (user) {
+            if (userHost) {
+                this.updateUserTokenAccess(user.id, token, userHost);
+            }
             return user;
         }
 
         return null;
     }
 
-    async generateUserToken(id) {
-        let token = tokens.generateUserToken();
+    async generateUserToken(userId, duration, comment, userHost) {
+        const token = tokens.generateUserToken();
+        const now = Helpers.now();
+        const expires = duration ? now + duration : 0;
         await this.db.dbUsers('user_tokens').insert({
-            user_id: id,
+            user_id: userId,
             token: token,
-            created_at: Helpers.now(),
+            created_at: now,
+            expires_at: expires,
+            accessed_at: now,
+            last_ip: userHost,
+            comment: comment || '',
         });
         return token;
     }
 
+    async updateUserToken(userId, token, duration, comment) {
+        const expires = duration ? Helpers.now() + duration : 0;
+        const updateObj = Object.create(null);
+
+        if (duration !== null) {
+            updateObj.expires_at = expires;
+        }
+        if (comment) {
+            updateObj.comment = comment;
+        }
+        if (!Object.keys(updateObj).length) {
+            return 0;
+        }
+
+        return await this.db.dbUsers('user_tokens')
+        .update(updateObj)
+        .where('user_id', userId)
+        .where('token', token);
+    }
+
+    async updateUserTokenAccess(userId, token, userHost) {
+        await this.db.dbUsers('user_tokens').update({
+            accessed_at: Helpers.now(),
+            last_ip: userHost,
```

Comment:
Since `userHost` is optional in some parts, should this be checking if the param is given before updating that column?

Reply:
I check userHost before calling updateUserTokenAccess